{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pysam\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {},
   "outputs": [],
   "source": [
    "fa = pysam.FastaFile('hg38.fa') "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'taaccctaaccctaaccctaaccctaaccctaa'"
      ]
     },
     "execution_count": 130,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fa.fetch(reference='chr1', start=10000, end= 10000+ 33)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "ref = pd.DataFrame({'chr': fa.references, 'length': fa.lengths})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>248956422</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr10</td>\n",
       "      <td>133797422</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr11</td>\n",
       "      <td>135086622</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr11_KI270721v1_random</td>\n",
       "      <td>100316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr12</td>\n",
       "      <td>133275309</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>450</th>\n",
       "      <td>chrUn_GL000216v2</td>\n",
       "      <td>176608</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>451</th>\n",
       "      <td>chrUn_GL000218v1</td>\n",
       "      <td>161147</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>452</th>\n",
       "      <td>chrX</td>\n",
       "      <td>156040895</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>453</th>\n",
       "      <td>chrY</td>\n",
       "      <td>57227415</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>454</th>\n",
       "      <td>chrY_KI270740v1_random</td>\n",
       "      <td>37240</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>455 rows Ã— 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                         chr     length\n",
       "0                       chr1  248956422\n",
       "1                      chr10  133797422\n",
       "2                      chr11  135086622\n",
       "3    chr11_KI270721v1_random     100316\n",
       "4                      chr12  133275309\n",
       "..                       ...        ...\n",
       "450         chrUn_GL000216v2     176608\n",
       "451         chrUn_GL000218v1     161147\n",
       "452                     chrX  156040895\n",
       "453                     chrY   57227415\n",
       "454   chrY_KI270740v1_random      37240\n",
       "\n",
       "[455 rows x 2 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ref"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "fa.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "import pysam\n",
    "import math\n",
    "import numpy as np\n",
    "from typing import NamedTuple\n",
    "import array\n",
    "import gzip\n",
    "\n",
    "\n",
    "CG_CONTEXT_FORWARD_HASH = {'CGA':'CG', 'CGT':'CG', 'CGC':'CG', 'CGG':'CG', 'CGN':'CG', # 5 CG\n",
    "                           'CAG':'CHG', 'CTG':'CHG', 'CCG':'CHG',          # 3 CHG\n",
    "                           'CAA':'CHH', 'CAT':'CHH', 'CAC':'CHH',          # 9 CHH\n",
    "                           'CTA':'CHH', 'CTT':'CHH', 'CTC':'CHH',\n",
    "                           'CCA':'CHH', 'CCT':'CHH', 'CCC':'CHH'\n",
    "                           }\n",
    "\n",
    "CG_CONTEXT_REVERSE_HASH = {'ACG':'CG', 'TCG':'CG', 'CCG':'CG', 'GCG':'CG', 'NCG':'CG', # 5 CG\n",
    "                           'CAG':'CHG', 'CTG':'CHG', 'CGG':'CHG',          # 3 CHG\n",
    "                           'AAG':'CHH', 'ATG':'CHH', 'AGG':'CHH',          # 9 CHH\n",
    "                           'TAG':'CHH', 'TTG':'CHH', 'TGG':'CHH',\n",
    "                           'GAG':'CHH', 'GTG':'CHH', 'GGG':'CHH'\n",
    "                           }\n",
    "\n",
    "DI_CONTEXT_REVERSE_HASH = {'AG':'CT', 'TG':'CA', 'CG':'CG', 'GG':'CC'}\n",
    "\n",
    "# parse 3-letter context, CG/CHG/CHH etc.\n",
    "\n",
    "class CGContext:\n",
    "    \n",
    "    def __init__(self, chr, pos, base, threeletter):\n",
    "        pass\n",
    "\n",
    "# interval must be involved in single chr\n",
    "class GenomicInterval(NamedTuple):\n",
    "    chr: str\n",
    "    chr_length: int\n",
    "    start: int\n",
    "    end: int\n",
    "\n",
    "# `ref` includes ref bases from `start`-2 to `end`-2\n",
    "class FaGenomicInterval(NamedTuple):\n",
    "    chr: str\n",
    "    start: int\n",
    "    end: int\n",
    "    bases: str\n",
    "\n",
    "\n",
    "class GenomicIntervalGenerator:\n",
    "    \n",
    "    def __init__(self, fa: pysam.FastaFile, \n",
    "                 chrs = \"all\",\n",
    "                 start: int = 0,\n",
    "                 end: int = math.inf,\n",
    "                 step: int = 10_000,\n",
    "                 ) -> None:\n",
    "        \n",
    "        self.chrs = fa.references\n",
    "        self.lens = fa.lengths\n",
    "        if chrs == \"all\":\n",
    "            self.chrs_selected = self.chrs\n",
    "        else:\n",
    "            self.chrs_selected = chrs\n",
    "        \n",
    "        # print(self.chrs_selected)\n",
    "        self.start = start\n",
    "        self.end = end\n",
    "        self.step = step\n",
    "\n",
    "        assert(step > 0 and start < end)\n",
    "        assert(len(self.chrs) > 0 and len(self.chrs) == len(self.lens))\n",
    "\n",
    "\n",
    "    def __repr__(self):\n",
    "        return f'GenomicIntervalGenerator({len(self.chrs)} contig(s) with step {self.step})'\n",
    "    \n",
    "    def __iter__(self) -> GenomicInterval:\n",
    "        for (chr, len) in zip(self.chrs, self.lens):\n",
    "            if chr not in self.chrs_selected:\n",
    "                continue\n",
    "\n",
    "            end2 = min(self.end, len)\n",
    "            start = self.start\n",
    "            end = start + self.step\n",
    "            while (True):\n",
    "                # end += self.step\n",
    "                if start < end2:\n",
    "                    end = min(end, end2)\n",
    "                    yield GenomicInterval(chr=chr, chr_length=len, start=start, end=end)\n",
    "                    start = end\n",
    "                    end += self.step\n",
    "                else:\n",
    "                    break\n",
    "\n",
    "class MyFastaFile(pysam.FastaFile):\n",
    "\n",
    "    def rich_fetch(self, intrv: GenomicInterval, padding:int=2) -> FaGenomicInterval:\n",
    "\n",
    "        bases = self.fetch(reference=intrv.chr, \n",
    "                           start=max(0, intrv.start-2), \n",
    "                           end=max(intrv.chr_length, intrv.end+2)\n",
    "                           ).upper()\n",
    "        \n",
    "        # padding N\n",
    "        if intrv.start < padding:\n",
    "            bases += \"N\"*(padding-intrv.start)\n",
    "        if intrv.end > intrv.chr_length-padding:\n",
    "            bases += \"N\"*(intrv.chr_length-intrv.end)\n",
    "\n",
    "        return bases\n",
    "\n",
    "\n",
    "class Coverage(NamedTuple):\n",
    "    watson: np.array\n",
    "    crick: np.array\n",
    "\n",
    "def reverse_read(read) -> bool:\n",
    "    return read.is_reverse\n",
    "\n",
    "def forward_read(read) -> bool:\n",
    "    return not read.is_reverse\n",
    "\n",
    "class MyAlignmentFile(pysam.AlignmentFile):\n",
    "\n",
    "    # def __init__(self, )\n",
    "\n",
    "    def Watson_Crick_coverage(self, intrv: GenomicInterval, quality_threshold: int) -> Coverage:\n",
    "        cov_watson = self.count_coverage(contig=intrv.chr, \n",
    "                                         start=intrv.start, \n",
    "                                         stop=intrv.end, \n",
    "                                         quality_threshold=quality_threshold,\n",
    "                                         read_callback=forward_read\n",
    "                                         )\n",
    "        cov_crick = self.count_coverage(contig=intrv.chr, \n",
    "                                         start=intrv.start, \n",
    "                                         stop=intrv.end, \n",
    "                                         quality_threshold=quality_threshold,\n",
    "                                         read_callback=reverse_read\n",
    "                                         )\n",
    "        return Coverage(watson=np.array(cov_watson), crick=np.array(cov_crick))\n",
    "\n",
    "\n",
    "\n",
    "def methylExtractor(params: NamedTuple) -> None:\n",
    "    \n",
    "    fa = MyFastaFile(params.fafile)\n",
    "    bam = MyAlignmentFile(params.bamfile, 'rb')\n",
    "\n",
    "\n",
    "    # outputs\n",
    "    outfile_atcg = gzip.open(params.out_atcg, 'wt')\n",
    "\n",
    "    intervals = iter(GenomicIntervalGenerator(fa, \n",
    "                                              chrs= \"chr1\", \n",
    "                                              start = 100000, \n",
    "                                              end = 100000+ 55,\n",
    "                                              step=10))\n",
    "\n",
    "    for intrv in intervals:\n",
    "\n",
    "        # context size\n",
    "        consize = params.context_size\n",
    "        # ref sequences\n",
    "        bases = fa.rich_fetch(intrv, padding=consize-1).upper()\n",
    "\n",
    "\n",
    "        # bam coverages\n",
    "        covs = bam.Watson_Crick_coverage(intrv, quality_threshold=params.quality_threshold)\n",
    "\n",
    "        cov_sum_W = np.sum(covs.watson, axis=0)\n",
    "        cov_sum_C = np.sum(covs.crick, axis=0)\n",
    "\n",
    "        chr = intrv.chr\n",
    "        # pos\n",
    "        for i in range(intrv.end-intrv.start):\n",
    "            if cov_sum_W[i]+cov_sum_C[i] == 0:\n",
    "                continue\n",
    "            \n",
    "\n",
    "            j = i+2\n",
    "            base = bases[j]\n",
    "            if base == 'C':\n",
    "                # CG/CHG/CHH\n",
    "                CG_context = CG_CONTEXT_FORWARD_HASH[bases[j:(j+consize)]]\n",
    "                # dinucleatide context CA/CT/...\n",
    "                dicontext = bases[j:(j+2)]\n",
    "            elif base == 'G':\n",
    "                CG_context = CG_CONTEXT_REVERSE_HASH[bases[(j-consize+1):(j+1)]]\n",
    "                dicontext = DI_CONTEXT_REVERSE_HASH[bases[(j-1):(j+1)]]\n",
    "            else:\n",
    "                CG_context = dicontext = \"--\"\n",
    "\n",
    "            # write files\n",
    "            outfile_atcg.write(f'{intrv.chr}\\t{base}\\t{intrv.start+i}\\t{CG_context}\\t{dicontext}\\t{covs.watson[0,i]}\\t{covs.watson[3,i]}\\t{covs.watson[1,i]}\\t{covs.watson[2,i]}\\t{covs.crick[0,i]}\\t{covs.crick[3,i]}\\t{covs.crick[1,i]}\\t{covs.crick[2,i]}\\n')\n",
    "\n",
    "    \n",
    "    # close file handers\n",
    "\n",
    "    fa.close()\n",
    "    bam.close()\n",
    "    outfile_atcg.close()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "class Parameters(NamedTuple):\n",
    "    fafile: str\n",
    "    bamfile: str\n",
    "    out_atcg: str\n",
    "    quality_threshold: int\n",
    "    context_size: int\n",
    "    \n",
    "params = Parameters(fafile='./hg38.fa', \n",
    "                    bamfile='./A549.bam',\n",
    "                    out_atcg='./A549.ATCGmap.gz',\n",
    "                    quality_threshold=15, # base seq quality\n",
    "                    context_size=3, # CHG etc.\n",
    "                    )\n",
    "methylExtractor(params)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<generator object GenomicIntervalGenerator.__iter__ at 0x7f9586f30260>"
      ]
     },
     "execution_count": 102,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "\n",
    "params = Parameters(fafile='./hg38.fa', bamfile='./A549.bam')\n",
    "\n",
    "fa = pysam.FastaFile(params.fafile)\n",
    "interval = iter(GenomicIntervalGenerator(fa, \n",
    "                                         chrs = [\"chr1\", \"chr2\"],\n",
    "                                         start= 100_000_001, \n",
    "                                         end = 200_000_002, \n",
    "                                         step=50_000_000)\n",
    "                                         )\n",
    "\n",
    "interval"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "next(interval)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "fa.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 141,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'N' in \"CNG\""
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "pysam",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
